clients/command-kafka/src/main/java/org/eclipse/hono/client/command/kafka/KafkaBasedCommandContext.java

```diff
@@ -128,7 +153,48 @@ private void reject(final int status, final String cause) {
         }
         final Span span = getTracingSpan();
         Tags.HTTP_STATUS.set(span, status);
-        span.finish();
+        if (isRequestResponseCommandWithCorrelationId()) {
+            final String nonNullCause = Optional.ofNullable(cause).orElse("Command message rejected");
+            sendDeliveryFailureCommandResponseMessage(status, nonNullCause, span)
+                    .onComplete(v -> span.finish());
+        } else {
+            logErrorIfCorrelationIdMissing(span);
+            span.finish();
+        }
+    }
+
+    private boolean isRequestResponseCommandWithCorrelationId() {
+        return !command.isOneWay() && getCorrelationId() != null;
+    }
+
+    private void logErrorIfCorrelationIdMissing(final Span span) {
+        if (!command.isOneWay() && getCorrelationId() == null) {
+            TracingHelper.logError(span, "can't send command response message - no correlation id set");
+        }
+    }
+
+    private String getCorrelationId() {
+        // extract correlation id from headers; command could be invalid in which case command.getCorrelationId() throws an exception
+        return KafkaRecordHelper
+                .getHeaderValue(command.getRecord().headers(), MessageHelper.SYS_PROPERTY_CORRELATION_ID, String.class)
+                .orElse(null);
+    }
+
+    private Future<Void> sendDeliveryFailureCommandResponseMessage(final int status, final String error, final Span span) {
+        final JsonObject payloadJson = new JsonObject();
+        payloadJson.put("error", error != null ? error : "");
+        final String correlationId = getCorrelationId();
+        final CommandResponse commandResponse = new CommandResponse(
+                command.getTenant(),
+                command.getDeviceId(),
+                payloadJson.toBuffer(),
+                CommandConstants.CONTENT_TYPE_DELIVERY_FAILURE_NOTIFICATION,
```

Comment:
IMHO we should document this content type in the Kafka C&C API.

Reply:
It's already part of the documentation, see https://github.com/eclipse/hono/pull/2712/files#diff-ad0c0c7fe66e910cb2d3b009825b6f8d9640e8ffac92f52480cb486b03b05e32R108.